### PR TITLE
[Persistence] Followup on PrePersistence deprecation & Block lifecycle 

### PR DESCRIPTION
**Pull request by Olshansk**
# Objective

Follow up on both critical and non-critical fixes to https://github.com/pokt-network/pocket/pull/140.

# Origin Document

https://github.com/pokt-network/pocket/pull/140 deprecated PrePersistence and implemented multiple contexts. This PR is a follow with some stylistic changes, improvements in the readability of the tests, the addition of unit tests, bug fixes in some SQL business logic, and implementation of the read context.

Followup work is being tracked in https://github.com/pokt-network/pocket/issues/149.

## Learnings to streamline workflow in the future
- Do not conflate non-critical stylistic changes as it leads to more conflicts down the road
- Implemented an automated "LocalNet" for easier testing
- Try to plan work ahead of time so we can work in smaller PRs / iterations

## Changelog / changes

**Main persistence module changes:**

- Split `ConnectAndInitializeDatabase` into `connectToDatabase` and `initializeDatabase`
  - This enables creating multiple contexts in parallel without re-initializing the DB connection
- Fix the SQL query used in `SelectActors`, `SelectAccounts` & `SelectPools`
  - Add a generalized unit test for all actors
- Remove `NewPersistenceModule` and an injected `Config` + `Create`
  - This improves isolation a a “injection-like” paradigm for unit testing
- Change `SetupPostgresDocker` to `SetupPostgresDockerPersistenceMod`
  - This enables more “functional” like testing by returning a persistence module and avoiding global testing variables
  - Only return once a connection to the DB has been initialized reducing the likelihood of test race conditions
- Implemented `NewReadContext` with a proper read-only context

**Secondary persistence module changes**

- Improve return values in `Commit` and `Release` (return error, add logging, etc…)
- Add `pgx.Conn` pointer to `PostgresDB`
- `s/db/conn/g` and `s/conn/tx/g` in some (not all) places where appropriate
- Make some exported variables / functions unexported for readability & access purposes
- Add a few helpers for persistence related unit testing
- Added unit tests and TODOs for handling multiple read/write contexts

## General milestone checklist
- [x] Update all the relevant CHANGELOGs
- [ ] Update all the relevant READMEs
- [ ] Update the source code tree explanation
- [ ] Add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [x] Create followup milestones + issues
- [x] Document small TODO along the way

## Follow-up Work

All follow-up work is being tracked in https://github.com/pokt-network/pocket/issues/149

---

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] If applicable, I have made corresponding changes to related or global README
- [ ] If applicable, I have added added new diagrams using [memaid.js](https://mermaid-js.github.io)
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: shared/tests/utility_module/block_test.go [hunk 1]
```diff
@@ -72,11 +79,14 @@ func TestUtilityContext_BeginBlock(t *testing.T) {
...
 		require.NoError(t, err)
 	}
+
+	// DISCUSS_IN_THIS_COMMIT: Why did we comment this part out?
 	// beginBlock logic verify
-	//missed, err := ctx.GetValidatorMissedBlocks(byzantine.Address) TODO not yet implemented
-	//require.NoError(t, err)
-	//require.False(t, missed != 1, fmt.Sprintf("wrong missed blocks amount; expected %v got %v", 1, byzantine.MissedBlocks))
-	ctx.Context.Release() // TODO (team) need a golang specific solution for teardown
+	// missed, err := ctx.GetValidatorMissedBlocks(byzantine.Address) TODO not yet implemented
```
Comment (andrewnguyen22) [inline shared/tests/utility_module/block_test.go:85]: See TODO (not yet implemented)

File: persistence/schema/shared_sql.go [hunk 2]
```diff
@@ -119,6 +121,8 @@ func Insert(
 							  height=EXCLUDED.height`,
 		tableName, actorSpecificParam,
 		actor.Address, actor.PublicKey, actor.StakedTokens, actorSpecificParamValue,
+		// DISCUSS_IN_THIS_COMMIT: Why did we change this?
```
Comment (andrewnguyen22) [inline persistence/schema/shared_sql.go:124]: I don't have a good answer for this, revert?

File: persistence/shared_sql.go [hunk 3]
```diff
@@ -4,13 +4,15 @@ import (
...
 	"fmt"
+
 	"github.com/jackc/pgx/v4"
 
 	"github.com/pokt-network/pocket/persistence/schema"
 	"github.com/pokt-network/pocket/shared/types"
 )
 
-// IMPROVE(team): Move this into a proto enum
+// IMPROVE(team): Move this into a proto enum. We are not using `iota` for the time being
```
Comment (andrewnguyen22) [inline persistence/shared_sql.go:14]: (Non-Actionable) StakeStatus is deprecated at the state level in #154 expect more cleanup to come here

File: persistence/test/account_test.go [hunk 4]
```diff
@@ -294,13 +303,18 @@ func newTestAccount(t *testing.T) typesGenesis.Account {
...
+}
+
 func newTestPool(t *testing.T) typesGenesis.Pool {
 	_, err := crypto.GenerateAddress()
 	if t != nil {
 		require.NoError(t, err)
 	}
 	return typesGenesis.Pool{
-		Name: DefaultPoolName,
+		Name: fmt.Sprintf("%s_%d", DefaultPoolName, rand.Int()),
```
Comment (andrewnguyen22) [inline persistence/test/account_test.go:317]: (Non-Actionable) Pools become proto-enum in #154

File: persistence/test/application_test.go [hunk 5]
```diff
@@ -246,6 +236,16 @@ func TestGetSetStakeAmount(t *testing.T) {
 	require.Equal(t, newStakeAmount, stakeAmountAfter, "unexpected status")
 }
 
+func TestGetAllApps(t *testing.T) {
+	db := NewTestPostgresContext(t, 0)
+
+	updateApp := func(db *persistence.PostgresContext, app *genesis.App) error {
+		return db.UpdateApp(app.Address, "https://olshansky.info", app.MaxRelays, []string{"OLSH"})
```
Comment (andrewnguyen22) [inline persistence/test/application_test.go:243]: NIT: Can make these constants and global var
`https://olshansky.info`
`[]string{"OLSH"}`

File: persistence/test/generic_test.go [hunk 6]
```diff
@@ -34,6 +36,92 @@ func NewTestGenericActor[T any](protocolActorSchema schema.ProtocolActorSchema,
...
+	actors, err = getAllActors(1)
+	require.NoError(t, err)
+	require.Len(t, actors, initialCount+2)
+
+	// Check height 10
+	actors, err = getAllActors(10)
+	require.NoError(t, err)
+	require.Len(t, actors, initialCount+3)
+
+	// DISCUSS_IN_THIS_COMMIT: Since we do not support `DeleteActor`, should we filter here based on status? If so, tests need to be updated.
```
Comment (andrewnguyen22) [inline persistence/test/generic_test.go:117]: (Non-Actionable) 'Delete' is in limbo for now. I say let's document/leave it

**Comment by andrewnguyen22** [comment]
First pass - here are my thoughts:
- Most of it is cleanup (not a bad thing)
- There will be a ton of conflicts with #154 (a bad thing)

Per comments in #128 let's talk offline about merge strategy. 

Could get quite messy given all of the outstanding things

File: persistence/shared_sql.go (hunk 3 above)
Comment (Olshansk) [reply persistence/shared_sql.go:14]: ACK

File: persistence/test/account_test.go (hunk 4 above)
Comment (Olshansk) [reply persistence/test/account_test.go:317]: ACK

File: persistence/test/application_test.go (hunk 5 above)
Comment (Olshansk) [reply persistence/test/application_test.go:243]: Any objections to naming it `OlshanskyURL` and `OlshanskyChain`

File: shared/tests/utility_module/block_test.go (hunk 1 above)
Comment (Olshansk) [reply shared/tests/utility_module/block_test.go:85]: Removed the `DISCUSS_IN_THIS_COMMIT`

File: persistence/test/generic_test.go (hunk 6 above)
Comment (Olshansk) [reply persistence/test/generic_test.go:117]: Comment updated.

**Comment by andrewnguyen22** [comment]
> EDIT: Moving this to https://github.com/pokt-network/pocket/pull/165.
>
> I'm updating this to use DISTINCT and avoiding the IN query. Even though this probably yields some performance gains (did not benchmark), I think it's more readable / understandable to see what query we're making.
>
> Also going to add some unit tests here too.

> EDIT: Moving this to https://github.com/pokt-network/pocket/pull/165.
> 
> I added a few unit tests and there seemed to be a bug in the previous approach but I didn't dive deep enough to understand the reasoning. Here is the new business logic:
> 
> func SelectAccounts(height int64, tableName string) string {
```
	return fmt.Sprintf(`
			SELECT DISTINCT ON (address) address, balance, height
			FROM %s
			WHERE height<=%d
			ORDER BY address, height DESC
       `, tableName, height)
}
```

The original approach
```
			SELECT address, balance, height
			FROM %s WHERE height<=%d AND (height,address) IN (SELECT MAX(height),address from %s GROUP BY address)
```

(attempts to) ensure that the latest height is selected from the set.

Can you point me to documentation that shows that `DISTINCT ON` will always select the `first row` from the `ORDER BY` thus guaranteeing that the latest height is always selected.

**Comment by Olshansk** [comment]
@andrewnguyen22 While this is still a WIP, can you please take a look at all the `DISCUSS_IN_THIS_COMMIT` comments?

If you pull the branch locally, you can list of all of in terminal using `make todo_this_commit` :)

File: persistence/test/application_test.go (hunk 5 above)
Comment (andrewnguyen22) [reply persistence/test/application_test.go:233]: Lol no objection

File: persistence/schema/shared_sql.go (hunk 2 above)
Comment (Olshansk) [reply persistence/schema/shared_sql.go:124]: I reverted it and it broke the unit tests. Going to investigate

**Comment by Olshansk** [comment]
@andrewnguyen22 This is still incomplete but I made some progress. Most of the remaining action items can be seen with `make todo_this_commit`.

File: persistence/schema/shared_sql.go (hunk 2 above)
Comment (Olshansk) [reply persistence/schema/shared_sql.go]: Fixed: https://github.com/pokt-network/pocket/pull/165/commits/7148c17685b28f0734139447c19550712305ea2f

File: persistence/test/application_test.go [hunk 7]
```diff
@@ -209,26 +190,25 @@ func newTestApp() (*typesGenesis.App, error) {
...
 		MaxRelays:       DefaultMaxRelays,
-		StakedTokens:    typesGenesis.DefaultStake,
+		StakedTokens:    genesis.DefaultStake,
 		PausedHeight:    DefaultPauseHeight,
 		UnstakingHeight: DefaultUnstakingHeight,
 		Output:          outputAddr,
 	}, nil
 }
 
+// DISCUSS_IN_THIS_COMMIT: We are only calling these functions and tests for apps, but need to
```
Comment (andrewnguyen22) [inline persistence/test/application_test.go:207]: Okay, is this a discussion?

File: persistence/test/setup_test.go [hunk 8]
```diff
@@ -40,38 +39,62 @@ var (
...
+func NewTestPostgresContext(t *testing.T, height int64) *persistence.PostgresContext {
+	ctx, err := testPersistenceModule.NewRWContext(height)
+	require.NoError(t, err)
+	db := &persistence.PostgresContext{
+		Height: height,
+		DB:     ctx.(persistence.PostgresContext).DB,
+	}
+	t.Cleanup(func() {
+		require.NoError(t, db.Release())
+		// DISCUSS_IN_THIS_COMMIT: Do we need this given that `Release` reverses the changes made to the database?
```
Comment (andrewnguyen22) [inline persistence/test/setup_test.go:65]: I think it's cleaner and safer IMO to restart the whole stack. But not too strong on it

File: persistence/test/setup_test.go [hunk 9]
```diff
@@ -40,38 +39,62 @@ var (
...
+	if err != nil {
+		log.Fatalf("Error creating new context: %s", err)
+	}
+	db := persistence.PostgresContext{
+		Height: height,
+		DB:     ctx.(persistence.PostgresContext).DB,
+	}
+	f.Cleanup(func() {
+		db.Release()
+		// DISCUSS_IN_THIS_COMMIT: Do we need this given that `Release` reverses the changes made to the database?
```
Comment (andrewnguyen22) [inline persistence/test/setup_test.go:84]: See comment above

File: utility/types/message.go [hunk 10]
```diff
@@ -18,6 +18,7 @@ import (
 validations possible before even checking the state storage layer.
 */
 
+// DISCUSS_IN_THIS_COMMIT: What's the best location for these shared constants?
```
Comment (andrewnguyen22) [inline utility/types/message.go:21]: Not sure

File: persistence/account.go [hunk 11]
```diff
@@ -10,6 +10,13 @@ import (
 
 // TODO(https://github.com/pokt-network/pocket/issues/102): Generalize Pool and Account operations.
 
+// DISCUSS_IN_THIS_COMMIT: if we want to make use of `defaultAccountAmountStr`, then we should
+//                         check if the error returned by the DB is `RowNotFound` and return a nil
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/account.go]: I'm using `pgx.ErrNoRows` for the check which I think is reasonable given that it is associated with the`PostgresContext` struct, so a different platform / struct would check for a different error.

Also added 2 unit test to verify this moving forward. [`3e54025` (#165)](https://github.com/pokt-network/pocket/pull/165/commits/3e54025dce6d0c10699ae35873316587142e145e)

File: persistence/db.go [hunk 12]
```diff
@@ -18,75 +18,74 @@ const (
...
 )
 
+var protocolActorSchemas = []schema.ProtocolActorSchema{
+	schema.ApplicationActor,
+	schema.FishermanActor,
+	schema.ServiceNodeActor,
+	schema.ValidatorActor,
+}
+
+// DISCUSS_IN_THIS_COMMIT: Do we still need this?
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/db.go:28]: The rand pacakge is small and we're using it elsewhere so 100% not the problem.

I'm just not sure what we're seeding for here in the persistence module. If anything, we can do it in `node.go`?

tl;dr Tests still passed after removing it, but don't object to adding it back in.

File: persistence/genesis.go [hunk 13]
```diff
@@ -25,67 +30,85 @@ func (pm *persistenceModule) PopulateGenesisState(state *genesis.GenesisState) {
...
+			log.Fatalf("an error occurred inserting a validator in the genesis state: %s", err.Error())
 		}
-		if err = addValueToPool(genesis.FishermanStakePoolName, act.StakedTokens); err != nil {
-			log.Fatal(fmt.Sprintf("an error occurred inserting staked tokens into %s pool", genesis.FishermanStakePoolName))
+		if err = addValueToPool(genesis.ValidatorStakePoolName, val.StakedTokens); err != nil {
+			log.Fatalf("an error occurred inserting staked tokens into %s pool", genesis.ValidatorStakePoolName)
+		}
+	}
+
+	// DISCUSS_IN_THIS_COMMIT: Do protocol actors need corresponding accounts?
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/genesis.go:75]: Let's sync on this offline.

File: persistence/module.go [hunk 14]
```diff
@@ -14,56 +15,50 @@ var _ modules.PersistenceModule = &persistenceModule{}
...
+
+	persistenceMod := &persistenceModule{
+		bus:         nil,
+		postgresURL: cfg.Persistence.PostgresUrl,
+		nodeSchema:  cfg.Persistence.NodeSchema,
+		blockStore:  blockStore,
+	}
+
+	// DISCUSS_IN_THIS_COMMIT: Is `Create` the appropriate location for this or should it be `Start`?
+	// DISCUSS_IN_THIS_COMMIT: Thoughts on bringing back `shouldHydrateGenesisDb`? It allowed LocalNet
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/module.go]: Sounds good. Let's ignore it in this commit.

File: persistence/module.go [hunk 15]
```diff
@@ -14,56 +15,50 @@ var _ modules.PersistenceModule = &persistenceModule{}
...
-	return pm, nil
+
+	persistenceMod := &persistenceModule{
+		bus:         nil,
+		postgresURL: cfg.Persistence.PostgresUrl,
+		nodeSchema:  cfg.Persistence.NodeSchema,
+		blockStore:  blockStore,
+	}
+
+	// DISCUSS_IN_THIS_COMMIT: Is `Create` the appropriate location for this or should it be `Start`?
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/module.go:47]: Leaving here to reduce PR scope and adding this comment:

```
	// TECHDEBT: reconsider if this is the best place to call `populateGenesisState`. Note that
	// this forces the genesis state to be reloaded on every node startup until state sync is
	// implemented.
```

File: persistence/test/application_test.go (hunk 7 above)
Comment (Olshansk) [reply persistence/test/application_test.go:207]: Sorry, this was mean to be a `TODO_IN_THIS_COMMIT`

File: persistence/test/setup_test.go (hunk 8 above)
Comment (Olshansk) [reply persistence/test/setup_test.go:65]: I'm going to remove it so that it's easier for us to "run into errors" as we add more changes with unexpected circumstances.

File: persistence/genesis.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/genesis.go:75]: tl;dr When an unstake happens, it is set to 0.

[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/genesis.go:75]: tl;dr When an unstake happens, the amount is set appropriately. Otherwise, it defaults to 0.

**Comment by Olshansk** [comment]
@andrewnguyen22 Please take a look at this PR and lmk if there any additional changes you'd like for me to make.

File: persistence/test/module_test.go [hunk 16]
```diff
@@ -1,35 +1,129 @@
...
+	require.Equal(t, contextBOriginalAmount, contextAOriginalAmount)
+}
+
+// TODO(pocket/issues/149): Need to add support for this sort of test. The call to
+// `contextB.SetAccountAmount` currently hangs because we have multiple writes contexts at the
+// same height. Some potential solutions may include:
+// - keeping a set of all the write contexts and panicking
+// - Adding timeouts to the write contexts
+// - ???
+func TestTwoWriteContextsSameHeight(t *testing.T) {
```
Comment (andrewnguyen22) [inline persistence/test/module_test.go:55]: These tests don't make sense to me. As discussed, we should never have two write contexts in parallel

File: persistence/gov.go [hunk 17]
```diff
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"encoding/hex"
+
```
Comment (andrewnguyen22) [inline persistence/gov.go:5]: Why the newline here?

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]
All looks good. 99% there. Just two comments

File: persistence/gov.go (hunk 17 above)
Comment (Olshansk) [reply persistence/gov.go:5]: Go imports are automatically formatted by my vscode Golang extension based on the "latest best practices".

For me it's an auto-edit when I do a cmd+save.

I looked online (https://stackoverflow.com/a/38714480/768439) and seems like the best practice that the linter implements is grouping by package (e.g. github vs std) separated by a new line.

Screen Shot 2022-08-17 at 11 38 54 AM

File: persistence/test/module_test.go (hunk 16 above)
Comment (Olshansk) [reply persistence/test/module_test.go]: Removed tests and am now erroring on multiple write contexts.

**Comment by Olshansk** [comment]
@andrewnguyen22 Please take a close look at the commits between HEAD and `f9ddc0954de45f09559d7812de6e8b8857e9dc2d`. All the unit tests are passing but I also had to make several block lifecycle modifications for LocalNet to succeed.

File: consensus/module.go [hunk 18]
```diff
@@ -172,8 +172,9 @@ func (m *consensusModule) loadPersistedState() error {
 	}
 
 	// TODO: Populate the rest of the state from the persistence module: validator set, quorum cert, last block hash, etc...
-	m.Height = uint64(latestHeight) + 1 // +1 because the height of the consensus module is where it is actively participating in consensus
-	m.appHash = string(appHash)
+	// m.Height = uint64(latestHeight) + 1 // +1 because the height of the consensus module is where it is actively participating in consensus
```
Comment (andrewnguyen22) [inline consensus/module.go:175]: Why is this left as comments... 

What's going on with the appHash? Just logging?

File: persistence/genesis.go [hunk 19]
```diff
@@ -2,21 +2,33 @@ package persistence
...
 )
 
-func (pm *persistenceModule) PopulateGenesisState(state *genesis.GenesisState) { // TODO (Andrew) genericize with actors interface once merged with #111
-	poolValues := make(map[string]*big.Int, 0)
+// TODO(Andrew): generalize with the `actors interface`` once merged with #111
+// WARNING: This function crashes the process if there is an error populating the genesis state.
+func (m *persistenceModule) populateGenesisState(state *genesis.GenesisState) {
+	log.Println("Populating genesis state...")
+
+	// HACK: This is needed to avoid block a previous from interfering with the genesis state hydration
```
Comment (andrewnguyen22) [inline persistence/genesis.go:18]: This is very unclear to me. Populate Genesis should only be called once

File: persistence/gov.go [hunk 20]
```diff
@@ -27,6 +27,7 @@ func (p PostgresContext) InitParams() error {
 	if err != nil {
 		return err
 	}
+
```
Comment (andrewnguyen22) [inline persistence/gov.go:30]: Why newline?

File: persistence/module.go [hunk 21]
```diff
@@ -95,18 +123,31 @@ func (m *persistenceModule) NewRWContext(height int64) (modules.PersistenceRWCon
...
 			Blockstore: m.blockStore,
 		},
 	}, nil
 }
-
-func (m *persistenceModule) NewReadContext(height int64) (modules.PersistenceReadContext, error) {
-	return m.NewRWContext(height)
-	// TODO (Team) this can be completely separate from rw context.
-	// It should access the db directly rather than using transactions
+func (m *persistenceModule) ResetContext() error {
```
Comment (andrewnguyen22) [inline persistence/module.go:132]: This appears to only be a test function and not part of the lifecycle? 

Please add TODO to make it a test func only

Comment (andrewnguyen22) [reply persistence/module.go:132]: There's nothing I see here stoping this from being at the context level

File: shared/modules/persistence_module.go [hunk 22]
```diff
@@ -9,6 +9,7 @@ type PersistenceModule interface {
 	Module
 	NewRWContext(height int64) (PersistenceRWContext, error)
 	NewReadContext(height int64) (PersistenceReadContext, error)
+	ResetContext() error
```
Comment (andrewnguyen22) [inline shared/modules/persistence_module.go:12]: TODO remove from interface once moved to test function and/or re-scope it at the context level

**Comment by andrewnguyen22** [comment]
`Utility context not nil when preparing a new block? Releasing for now but should not happen`

File: persistence/genesis.go (hunk 19 above)
Comment (andrewnguyen22) [reply persistence/genesis.go:18]: This is non-actionable, but this comment speaks to how confusing the entire lifecycle issue is to me. I am not clear on the bug nor the solution at this time

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

**Comment by Olshansk** [comment]
@andrewnguyen22  comment regarding the genesis / lifecycle piece. 

When we added the blockstore in https://github.com/pokt-network/pocket/pull/121 , we also added a `shouldHydrateGenesis` function so that:
1. We don't run `populateGenesis` if there is already a previous state
2. Pick up the previous state if we already had one (hence the logic around loading the previous hash and height which I'm just logging)

https://github.com/pokt-network/pocket/pull/140/files?file-filters%5B%5D=.go&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.mod&file-filters%5B%5D=.sum&file-filters%5B%5D=No+extension&show-viewed-files=true&show-deleted-files=false

Screen Shot 2022-08-18 at 6 45 18 AM

My understanding is that per the other comments / discussions (screenshot below), we didn't want to bring it back in favor of the future configs changes happening, but I had to add a couple stop-gap hacks in the meantime so that we load up the genesis correctly every time.

I'd personally prefer to add it back in this PR so there are less hacks / confusion, but also unsure per you your other comments. Lmk how you think I should proceed.

Screen Shot 2022-08-18 at 6 46 21 AM

File: consensus/module.go (hunk 18 above)
Comment (Olshansk) [reply consensus/module.go:175]: Adding back w/ previous state hydration.

File: shared/modules/persistence_module.go (hunk 22 above)
Comment (Olshansk) [reply shared/modules/persistence_module.go:12]: Added.

**Review by andrewnguyen22 — APPROVED** [review]
